### PR TITLE
autoinstallable unimodules

### PR DESCRIPTION
**Pull request by tsapeta**
This is just a beginning of my work on a proof of concept for autoinstallable modules. 

It assumes that unimodules have `unimodule.json` file at the root folder, which contains configurations that already exist in `xdl` + some other stuff that might be required for Flutter/NativeScript.

Related PR: https://github.com/expo/expo/pull/3578

### CODE REVIEW

File: unimodule-config-template.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "libName": "expo-file-system",
```
Comment (brentvatne) [inline unimodule-config-template.json]: packageName or just package maybe

**Comment by fson** [comment]
Could this be a field in `package.json` (as opposed to a separate file)? This would avoid the need to redefine the package name in another file and in general cluttering the project root with less files is nice.

**Comment by brentvatne** [comment]
gave some feedback on slack and i think we're on the same page, this looks awesome and excited to get it in :)

File: unimodule-config-template.json (hunk 1 above)
Comment (ide) [reply unimodule-config-template.json]: or "name"?

**Comment by ide** [comment]
I also agree with @fson's comment. I think it'd make unimodules feel more lightweight if you could just add `"unimodule": true` (or `"unimodule": { ...configuration }` for more advanced cases) to your package.json file. The separate package name adds extra work and confusion (should it match package.json? can it be different?).

**Comment by tsapeta** [comment]
I'm aware that having separate file for unimodule config is not the perfect choice, but it makes searching for unimodule packages much easier as we can use glob pattern. If we move this field to `package.json` we would have to read all these files which could slow down the script. 
Also, `unimodule.json` would be more universal than `package.json` since in Flutter there is no `package.json`.

File: cocoapods.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  modules_paths.each { |module_path|
+    glob_pattern = File.join(module_path, '**', 'unimodule.json')
+
+    Dir.glob(glob_pattern) { |module_config_path|
+      unimodule_json = JSON.parse(File.read(module_config_path))
+      directory = File.dirname(module_config_path)
+      platforms = unimodule_json['platforms'] || ['ios']
+      targets = unimodule_json['targets'] || ['react-native']
+
+      if unimodule_supports_platform(platforms, 'ios') && unimodule_supports_target(targets, target) then
```
Comment (sjchmiela) [inline cocoapods.rb]: `then` is unnecessary and kinda verbose 🤷‍♂️

File: cocoapods.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        package_name = unimodule_json['name'] || get_package_name(directory)
+
+        if !exclude.include?(package_name) then
+          unimodule_config = { "subdirectory" => 'ios' }.merge(unimodule_json.fetch('ios', {}))
```
Comment (sjchmiela) [inline cocoapods.rb]:
```suggestion
          unimodule_config = { 'subdirectory' => 'ios' }.merge(unimodule_json.fetch('ios', {}))
```

File: cocoapods.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+def use_unimodules!(custom_options = {})
+  options = {
+    modules_paths: ['../node_modules'],
+    target: 'react-native',
+    exclude: [],
+  }.deep_merge(custom_options)
+
+  modules_paths = options.fetch(:modules_paths)
+  exclude = options.fetch(:exclude)
```
Comment (sjchmiela) [inline cocoapods.rb]:
```suggestion
  modules_to_exclude = options.fetch(:exclude)
```

File: gradle.groovy [hunk 5]
```diff
@@ -0,0 +1,102 @@
+import groovy.json.JsonSlurper
+
+def doesUnimoduleSupportPlatform(Map unimoduleJson, String platform) {
+  def platforms = unimoduleJson.platforms// || ['android']
```
Comment (sjchmiela) [inline gradle.groovy]:
```suggestion
  def platforms = unimoduleJson.platforms // || ['android']
```

File: gradle.groovy [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    exclude: [],
+  ] << customOptions
+
+  def unimodules = findUnimodules(options.target, options.exclude, options.modulesPaths)
+
+  for (unimodule in unimodules) {
+    if (options.configuration == 'expendency') {
+      expendency(unimodule.name)
+    } else {
+      Object dependency = project.project(":${unimodule.name}")
```
Comment (sjchmiela) [inline gradle.groovy]: This loosely typed `Object dependency` looks strange, maybe it would look better as the method argument? 🤔

**Comment by brentvatne** [comment]
@sjchmiela - from conversation on slack (sorry I should have just commented here)

screen shot 2019-02-26 at 10 05 43 am

File: gradle.groovy [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    }
+  }
+  return [
+    unimodules: unimodules.collect { entry -> entry.value },
+    duplicates: unimodulesDuplicates.unique()
+  ]
+}
+
+class Colors {
+  static final String NORMAL          = "\u001B[0m"
```
Comment (sjchmiela) [inline gradle.groovy]:
```suggestion
  static final String  NORMAL          = "\u001B[0m"
```

File: gradle.groovy [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+    for (moduleConfigPath in moduleConfigPaths) {
+      def unimoduleConfig = new File(moduleConfigPath)
+      def unimoduleJson = new JsonSlurper().parseText(unimoduleConfig.text)
+      def directory = unimoduleConfig.getParent()
+      
+      if (doesUnimoduleSupportPlatform(unimoduleJson, 'android') && doesUnimoduleSupportTarget(unimoduleJson, target)) {
+        def packageJsonFile = new File(directory, 'package.json')
+        def packageJson = new JsonSlurper().parseText(packageJsonFile.text)
+        def unimoduleName = unimoduleJson.name ? unimoduleJson.name : packageJson.name
```
Comment (ide) [inline gradle.groovy]:
```suggestion
        def unimoduleName = unimoduleJson.name ?: packageJson.name
```

**Comment by Pines-Cheng** [comment]
I can't wait to try it out. Is there any documentation on how to use it?

**Comment by brentvatne** [comment]
@Pines-Cheng - yeah it's been released for a while. this might be helpful https://docs.expo.io/versions/v32.0.0/bare/hello-world/ -- or just the readme for this project

**Pull request merged**